File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-2.js
class Solution {
    /**
     * @param {number[][]} heights
     * @return {number[][]}
     */
    pacificAtlantic(heights) {
        let ROWS = heights.length ;
        let COLS = heights[0].length ;
        let pacific =  new Set();
        let atlantic = new Set();
        let directions = [[0,1],[0,-1],[1,0],[-1,0]] ;
        function dfs(r,c,visited){
            visited.add(`${r},${c}`);
            for(let [dx,dy] of directions){
                let i = r+dx ;
                let j = c+dy ;
                if(i>=0 && j>=0 && i<ROWS && j<COLS && !visited.has(`${i},${j}`) && heights[i][j] >= heights[r][c]){
                    dfs(i,j,visited);
                }
            }
        }
        for(let i=0 ; i<ROWS ; i++){
            dfs(i,0,pacific);
            dfs(i,COLS-1,atlantic);
        }
        for(let j=0;j<COLS;j++){
            dfs(0,j,pacific);
            dfs(ROWS-1,j,atlantic);
        }
        let res=[] ;
        for(let i=0 ; i<ROWS ;i++){
            for(let j=0 ;j<COLS ;j++){
                if(pacific.has(`${i},${j}`) && atlantic.has(`${i},${j}`))
                   res.push([i,j]);
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/surrounded-regions/submission-2.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        directions=[(1,0),(-1,0),(0,1),(0,-1)]
        ROWS,COLS = len(board),len(board[0])
        visited=[[False for _ in range(COLS)]for _ in range(ROWS)]

        def dfs(i,j) :
            visited[i][j]=True
            for dx,dy in directions :
                x,y=i+dx,j+dy
                if x<0 or y<0 or x>=ROWS or y>=COLS :
                    visited[i][j]=False
                    return False
                if board[x][y]=="O" and not visited[x][y] :
                    if not dfs(x,y) :
                        visited[i][j] = False
                        return False
          #  board[i][j]="X"
            return True

        for i in range(ROWS) :
            for j in range(COLS) :
                if board[i][j]=="O" :
                    if dfs(i,j) :
                        board[i][j]="X"
        